libraries/lib-mixer/MixerSource.cpp

```diff
@@ -0,0 +1,469 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file MixerSource.cpp
+
+  Dominic Mazzoni
+  Markus Meyer
+  Vaughan Johnson
+
+  Paul Licameli split from Mix.cpp
+
+**********************************************************************/
+#include "MixerSource.h"
+
+#include "AudioGraphBuffers.h"
+#include "Envelope.h"
+#include "SampleTrackCache.h"
+#include "Resample.h"
+#include "WideSampleSequence.h"
+#include "float_cast.h"
+
+namespace {
+template<typename T, typename F> std::vector<T>
+initVector(size_t dim1, const F &f)
+{
+   std::vector<T> result( dim1 );
+   for (auto &row : result)
+      f(row);
+   return result;
+}
+
+template<typename T> std::vector<std::vector<T>>
+initVector(size_t dim1, size_t dim2)
+{
+   return initVector<std::vector<T>>(dim1,
+      [dim2](auto &row){ row.resize(dim2); });
+}
+}
+
+void MixerSource::MakeResamplers()
+{
+   for (size_t j = 0; j < mnChannels; ++j)
+      mResample[j] = std::make_unique<Resample>(
+         mResampleParameters.mHighQuality,
+         mResampleParameters.mMinFactor, mResampleParameters.mMaxFactor);
+}
+
+namespace {
+   //Note: The meaning of this function has changed (December 2012)
+   //Previously this function did something that was close to the opposite (but not entirely accurate).
+   /** @brief Compute the integral warp factor between two non-warped time points
+    *
+    * Calculate the relative length increase of the chosen segment from the original sound.
+    * So if this time track has a low value (i.e. makes the sound slower), the NEW warped
+    * sound will be *longer* than the original sound, so the return value of this function
+    * is larger.
+    * @param t0 The starting time to calculate from
+    * @param t1 The ending time to calculate to
+    * @return The relative length increase of the chosen segment from the original sound.
+    */
+double ComputeWarpFactor(const Envelope &env, double t0, double t1)
+{
+   return env.AverageOfInverse(t0, t1);
+}
+
+}
+
+size_t MixerSource::MixVariableRates(
+   unsigned nChannels, const size_t maxOut, float *floatBuffers[])
+{
+   const auto &[mT0, mT1, mSpeed, _] = *mTimesAndSpeed;
+   const bool backwards = (mT1 < mT0);
+
+   const auto seq = mInputSequence->GetSequence().get();
+   const double sequenceRate = seq->GetRate();
+   const double initialWarp = mRate / mSpeed / sequenceRate;
+   const double tstep = 1.0 / sequenceRate;
+   const auto sampleSize = SAMPLE_SIZE(floatSample);
+   // Find the last sample
+   const auto endPos = [&seq, mT1 = mT1, backwards]{
+      double endTime = seq->GetSequenceEndTime();
+      double startTime = seq->GetSequenceStartTime();
+      const double tEnd = backwards
+         ? std::max(startTime, mT1)
+         : std::min(endTime, mT1);
+      return seq->TimeToLongSamples(tEnd);
+   }();
+
+   auto pos = mSamplePos;
+   auto queueStart = mQueueStart;
+   auto queueLen = mQueueLen;
```

Comment:
This code was older than my SampleTrackCache, and that cache was also reused in spectrogram drawing.

The extra layer of queue is still useful to handle the mismatch of the number of samples fetched from the track, and the number of samples really produced, because of resampling.

I assume there is nothing for me to change in response because we will remove that cache anyway.

Reply:
> I assume there is nothing for me to change in response because we will remove that cache anyway.

Right, this is just a question for my understanding in the prospect of the caching refactoring, not a request for change.

I see where it comes from now, given the layer of resampling, thanks for the explanation.

When you say "remove that cache", you are talking about the `SampleTrackCache`, right?

If so, this caching for resampling is still a problem to time stretching, since it requests 65536 samples at once. This isn't as bad as the megabyte currently queried by `SampleTrackCache`, but still this will likely be a performance bottleneck. How to solve this?

Reducing its size? I suppose the queue must have size at least as large as the large playback buffer size possible, which might be tricky to find out.

The `StretchedWideSampleSequence` must solve the same problem : for a given requested output length, it doesn't know in advance how many samples must be pulled from its source. This makes me wonder if a good approach would be yet another, resampling decorator. Then `MixerSource` would need only one simpler method as opposed to currently two complex ones (`MixSameRate` and `MixVariableRates`). The sequence would then go `MixerSource > ResamplingSequence > StretchingSequence > WaveTrack` (`StretchingSequence` could also do the caching as proposed in my draft, or we'd pass it a `CachingSequence`, whichever.)

Reply:
Another interesting aspect to the resampling decorator: it could be freely injected at some other stage of the processing chain than the one proposed. I guess it's best to have it after stretching, yet who knows, there may be surprises along the way.